File: src/main/go/leetcode/lists/offerII/Offer46.go
package offerII

/**
 * @author wujingxinit@outlook.com
 * @date 2023/1/16 17:08
 */

func rightSideView(root *TreeNode) []int {

	if root == nil {
		return []int{}
	}

	q := make([]*TreeNode, 0)
	q = append(q, root)
	list := make([]int, 0)
	for len(q) > 0 {
		size := len(q)
		for size > 0 {
			node := q[0]
			q = q[1:]
			if size == 1 {
				list = append(list, node.Val)
			}
			if node.Left != nil {
				q = append(q, node.Left)
			}
			if node.Right != nil {
				q = append(q, node.Right)
			}
			size--
		}
	}
	return list
}

File: src/main/java/leetcode/lists/offerII/Offer46.java
package leetcode.lists.offerII;

import common.TreeNode;

import java.util.ArrayList;
import java.util.List;

/**
 * @author wujingxinit@outlook.com
 * @date 2023/7/25 11:07
 */
public class Offer46 {

    int high = 0;

    public List<Integer> rightSideView(TreeNode root) {
        List<Integer> list = new ArrayList<>();
        if (root == null) {
            return list;
        }
        dfs(root, list, 0);
        return list;
    }

    public void dfs(TreeNode root, List<Integer> list, int h) {
        if (h == high) {
            high++;
            list.add(root.val);
        }
        if (root.right != null) {
            dfs(root.right, list, h + 1);
        }
        if (root.left != null) {
            dfs(root.left, list, h + 1);
        }
    }
}
